File: Anul II/Medii de proiectare si programare/Laborator/lab08/C#/Ubb.BikeContest.Networking/ProtocolBuffers/ProtobufServicesProxy.cs
using Com.Ubb.Protocolbuffers;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Sockets;
using System.Text;
using System.Threading.Tasks;
using Ubb.BikeContest.Model;
using Ubb.BikeContest.Model.data;
using Ubb.BikeContest.Services;
using Google.Protobuf;

using Participant = Ubb.BikeContest.Model.Participant;
using Team = Ubb.BikeContest.Model.Team;
using Race = Ubb.BikeContest.Model.Race;
using User = Ubb.BikeContest.Model.User;
using RaceEntry = Ubb.BikeContest.Model.RaceEntry;

namespace Ubb.BikeContest.Networking.ProtocolBuffers
{
    public class ProtobufServicesProxy : IContestServices
    {
        private readonly string _host;
        private readonly int _port;
        private IMainObserver _client;
        private NetworkStream _stream;
        private TcpClient _connection;

        private Queue<BikeContestResponse> _responses;
        private volatile bool _finished;
        private EventWaitHandle _waitHandle;

        public ProtobufServicesProxy(string host, int port)
        {
            _host = host;
            _port = port;
            _responses = new Queue<BikeContestResponse>();
        }

        public List<Participant> FindAllParticipants()
        {
            BikeContestRequest request = ProtocolBuilderUtils.CreateGetParticipantsRequest();
            SendRequest(request);

            BikeContestResponse response = ReadResponse();
            if (response.Type == BikeContestResponse.Types.Type.Error)
            {
                throw new Exception(ProtocolBuilderUtils.GetErrorMessage(response));
            }

            return ProtocolBuilderUtils.GetParticipants(response);
        }

        public List<Team> FindAllTeams()
        {
            BikeContestRequest request = ProtocolBuilderUtils.CreateGetTeamsRequest();
            SendRequest(request);

            BikeContestResponse response = ReadResponse();
            if (response.Type == BikeContestResponse.Types.Type.Error)
            {
                throw new Exception(ProtocolBuilderUtils.GetErrorMessage(response));
            }

            return ProtocolBuilderUtils.GetTeams(response);
        }

        public List<Participant> GetParticipantsByTeam(long id)
        {
            BikeContestRequest request = ProtocolBuilderUtils.CreateGetParticipantsByTeamRequest(id);
            SendRequest(request);

            BikeContestResponse response = ReadResponse();
            if (response.Type == BikeContestResponse.Types.Type.Error)
            {
                throw new Exception(ProtocolBuilderUtils.GetErrorMessage(response));
            }

            return ProtocolBuilderUtils.GetParticipants(response);
        }

        public List<Race> GetRacesWhereNotRegisteredAndEngineCapacity(long participantId, int engineCapacity)
        {
            BikeContestRequest request 
                = ProtocolBuilderUtils.CreateGetUnregisteredRacesRequest(participantId, engineCapacity);
            SendRequest(request);

            BikeContestResponse response = ReadResponse();
            if (response.Type == BikeContestResponse.Types.Type.Error)
            {
                throw new Exception(ProtocolBuilderUtils.GetErrorMessage(response));
            }

            return ProtocolBuilderUtils.GetRaces(response);
        }

        public List<RaceDto> GetRacesWithParticipantCount()
        {
            BikeContestRequest request
                = ProtocolBuilderUtils.CreateGetRacesWithParticipantCountRequest();
            SendRequest(request);

            BikeContestResponse response = ReadResponse();
            if (response.Type == BikeContestResponse.Types.Type.Error)
            {
                throw new Exception(ProtocolBuilderUtils.GetErrorMessage(response));
            }

            return ProtocolBuilderUtils.GetRaceDtos(response);
        }

        public User Login(string username, string passwordToken, IMainObserver client)
        {
            InitialiseConnection();

            BikeContestRequest request = ProtocolBuilderUtils.CreateLoginRequest(username, passwordToken);
            SendRequest(request);

            BikeContestResponse response = ReadResponse();
            if (response.Type == BikeContestResponse.Types.Type.Error)
            {
                CloseConnection();
                throw new Exception(ProtocolBuilderUtils.GetErrorMessage(response));
            }
            else if (response.Type == BikeContestResponse.Types.Type.Ok)
            {
                this._client = client;
                return ProtocolBuilderUtils.GetUser(response);
            }
            else
            {
                throw new Exception("Error while reading response: expected Ok, got " + response.Type);
            }
        }

        public void Logout(User user, IMainObserver client)
        {
            BikeContestRequest request = ProtocolBuilderUtils.CreateLogoutRequest(user);
            SendRequest(request);

            BikeContestResponse response = ReadResponse();
            CloseConnection();

            if (response.Type == BikeContestResponse.Types.Type.Error)
            {
                throw new Exception(ProtocolBuilderUtils.GetErrorMessage(response));
            }
        }

        public void SaveParticipant(Participant newEntity)
        {
            BikeContestRequest request = ProtocolBuilderUtils.CreateNewParticipantRequest(newEntity);
            SendRequest(request);

            BikeContestResponse response = ReadResponse();

            if (response.Type == BikeContestResponse.Types.Type.Error)
            {
                throw new Exception(ProtocolBuilderUtils.GetErrorMessage(response));
            }
        }

        public void SaveRaceEntries(List<RaceEntry> newEntities)
        {
            BikeContestRequest request = ProtocolBuilderUtils.CreateNewRaceEntriesRequest(newEntities);
            SendRequest(request);

            BikeContestResponse response = ReadResponse();

            if (response.Type == BikeContestResponse.Types.Type.Error)
            {
                throw new Exception(ProtocolBuilderUtils.GetErrorMessage(response));
            }
        }

        private void SendRequest(BikeContestRequest request)
        {
            Console.WriteLine("Sending request..." + request);
            request.WriteDelimitedTo(_stream);
            _stream.Flush();
            Console.WriteLine("Request sent.");
        }

        private BikeContestResponse ReadResponse()
        {
            BikeContestResponse response = null;
            try
            {
                _waitHandle.WaitOne();
                lock (_responses)
                {
                    response = _responses.Dequeue();
                }
            }
            catch (Exception e) 
            {
                Console.WriteLine(e.StackTrace);
            }
            return response;
        }

        private void InitialiseConnection()
        {
            try
            {
                _connection = new TcpClient(_host, _port);
                _stream = _connection.GetStream();
                _finished = false;
                _waitHandle = new AutoResetEvent(false);
                StartReader();
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        private void CloseConnection()
        {
            _finished = true;
            try
            {
                _stream.Close();
                _connection.Close();
                _waitHandle.Close();
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        private void StartReader()
        {
            Thread worker = new Thread(Run);
            worker.Start();
        }

        private bool IsUpdateResponse(BikeContestResponse response)
        {
            return response.Type == BikeContestResponse.Types.Type.NewParticipant
                || response.Type == BikeContestResponse.Types.Type.NewRace
                || response.Type == BikeContestResponse.Types.Type.NewTeam
                || response.Type == BikeContestResponse.Types.Type.UpdatedRaces;
        }

        private void HandleUpdate(BikeContestResponse response)
        {
            if (IsUpdateResponse(response))
            {
                try
                {
                    switch (response.Type)
                    {
                        case BikeContestResponse.Types.Type.NewParticipant:
                            _client.ParticipantAdded(ProtocolBuilderUtils.GetParticipant(response));
                            break;
                        case BikeContestResponse.Types.Type.NewRace:
                            break;
                        case BikeContestResponse.Types.Type.NewTeam:
                            break;
                        case BikeContestResponse.Types.Type.UpdatedRaces:
                            _client.RaceEntriesAdded(ProtocolBuilderUtils.GetRaceDtos(response));
                            break;
                    }
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }
            }
        }

        public virtual void Run()
        {
            while (!_finished)
            {
                try
                {
                    BikeContestResponse response = BikeContestResponse.Parser.ParseDelimitedFrom(_stream);
                    Console.WriteLine("Response received: " + response);

                    if (IsUpdateResponse(response))
                    {
                        HandleUpdate(response);
                    }
                    else
                    {
                        lock (_responses)
                        {
                            _responses.Enqueue(response);
                        }
                        _waitHandle.Set();
                    }
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }
            }
        }

        Participant IContestServices.GetParticipantByData(Participant participant)
        {
            BikeContestRequest request
                = ProtocolBuilderUtils.CreateGetParticipantByDataRequest(participant);
            SendRequest(request);

            BikeContestResponse response = ReadResponse();
            if (response.Type == BikeContestResponse.Types.Type.Error)
            {
                throw new Exception(ProtocolBuilderUtils.GetErrorMessage(response));
            }

            return ProtocolBuilderUtils.GetParticipant(response);
        }

        List<RaceEntry> IContestServices.GetEntriesByRace(long raceId)
        {
            BikeContestRequest request
                = ProtocolBuilderUtils.CreateGetEntriesByRaceRequest(raceId);
            SendRequest(request);

            BikeContestResponse response = ReadResponse();
            if (response.Type == BikeContestResponse.Types.Type.Error)
            {
                throw new Exception(ProtocolBuilderUtils.GetErrorMessage(response));
            }

            return ProtocolBuilderUtils.GetRaceEntries(response);
        }

        Race IContestServices.GetRaceByName(string name)
        {
            BikeContestRequest request
                = ProtocolBuilderUtils.CreateGetRaceByNameRequest(name);
            SendRequest(request);

            BikeContestResponse response = ReadResponse();
            if (response.Type == BikeContestResponse.Types.Type.Error)
            {
                throw new Exception(ProtocolBuilderUtils.GetErrorMessage(response));
            }

            return ProtocolBuilderUtils.GetRace(response);
        }

        List<Race> IContestServices.GetRacesByEngineCapacity(int engineCapacity)
        {
            BikeContestRequest request
                = ProtocolBuilderUtils.CreateGetRacesByEngineCapacityRequest(engineCapacity);
            SendRequest(request);

            BikeContestResponse response = ReadResponse();
            if (response.Type == BikeContestResponse.Types.Type.Error)
            {
                throw new Exception(ProtocolBuilderUtils.GetErrorMessage(response));
            }

            return ProtocolBuilderUtils.GetRaces(response);
        }

        Team IContestServices.GetTeamByName(string name)
        {
            BikeContestRequest request
                = ProtocolBuilderUtils.CreateGetTeamByNameRequest(name);
            SendRequest(request);

            BikeContestResponse response = ReadResponse();
            if (response.Type == BikeContestResponse.Types.Type.Error)
            {
                throw new Exception(ProtocolBuilderUtils.GetErrorMessage(response));
            }

            return ProtocolBuilderUtils.GetTeam(response);
        }

        void IContestServices.SaveRace(Race race)
        {
            throw new NotImplementedException();
        }
    }
}

File: Anul II/Medii de proiectare si programare/Laborator/lab08/Java/networking/src/main/java/com/ubb/protocolbuffers/ProtobufServicesProxy.java
package com.ubb.protocolbuffers;

import com.ubb.IContestServices;
import com.ubb.IMainObserver;
import com.ubb.dto.RaceDTO;
import com.ubb.exceptions.ContestDataException;
import com.ubb.model.*;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.util.Collection;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

public class ProtobufServicesProxy implements IContestServices {
    private final String host;
    private final int port;
    private IMainObserver client;
    private InputStream inputStream;
    private OutputStream outputStream;
    private Socket connection;
    private final BlockingQueue<BikeContestProtocol.BikeContestResponse> responses;
    private volatile boolean finished;

    public ProtobufServicesProxy (String host, int port) {
        this.host = host;
        this.port = port;
        this.responses = new LinkedBlockingQueue<>();
    }

    @Override
    public User login(String username, String passwordToken, IMainObserver client) throws ContestDataException {
        initialiseConnection();
        BikeContestProtocol.BikeContestRequest request
                = ProtocolBuilderUtils.createLoginRequest(username, passwordToken);
        sendRequest(request);

        BikeContestProtocol.BikeContestResponse response = readResponse();
        if (response.getType() == BikeContestProtocol.BikeContestResponse.Type.Error) {
            closeConnection();
            throw new ContestDataException(ProtocolBuilderUtils.getErrorMessage(response));
        } else if (response.getType() == BikeContestProtocol.BikeContestResponse.Type.Ok) {
            this.client = client;
            return ProtocolBuilderUtils.getUser(response);
        } else {
            throw new ContestDataException("Authentication failed : expected OK response, got " + response.getType());
        }
    }

    @Override
    public void logout(User user, IMainObserver client) throws ContestDataException {
        BikeContestProtocol.BikeContestRequest request = ProtocolBuilderUtils.createLogoutRequest(user);
        sendRequest(request);

        BikeContestProtocol.BikeContestResponse response = readResponse();
        closeConnection();

        if (response.getType() == BikeContestProtocol.BikeContestResponse.Type.Error) {
            throw new ContestDataException(ProtocolBuilderUtils.getErrorMessage(response));
        }
    }

    @Override
    public Collection<Participant> getParticipantsByTeam(Long teamID) throws ContestDataException {
        BikeContestProtocol.BikeContestRequest request
                = ProtocolBuilderUtils.createGetParticipantsByTeamRequest(teamID);
        sendRequest(request);

        BikeContestProtocol.BikeContestResponse response = readResponse();
        if (response.getType() == BikeContestProtocol.BikeContestResponse.Type.Error) {
            throw new ContestDataException(ProtocolBuilderUtils.getErrorMessage(response));
        }
        return ProtocolBuilderUtils.getParticipants(response);
    }

    @Override
    public Optional<Participant> getParticipantByData(Participant participant) throws ContestDataException {
        BikeContestProtocol.BikeContestRequest request
                = ProtocolBuilderUtils.createGetParticipantByDataRequest(participant);
        sendRequest(request);

        BikeContestProtocol.BikeContestResponse response = readResponse();
        if (response.getType() == BikeContestProtocol.BikeContestResponse.Type.Error) {
            throw new ContestDataException(ProtocolBuilderUtils.getErrorMessage(response));
        }
        return Optional.of(ProtocolBuilderUtils.getParticipant(response));
    }

    @Override
    public Optional<Race> getRaceByName(String name) throws ContestDataException {
        BikeContestProtocol.BikeContestRequest request
                = ProtocolBuilderUtils.createGetRaceByNameRequest(name);
        sendRequest(request);

        BikeContestProtocol.BikeContestResponse response = readResponse();
        if (response.getType() == BikeContestProtocol.BikeContestResponse.Type.Error) {
            throw new ContestDataException(ProtocolBuilderUtils.getErrorMessage(response));
        }
        return Optional.of(ProtocolBuilderUtils.getRace(response));
    }

    @Override
    public Collection<RaceEntry> getEntriesByRace(Long raceID) throws ContestDataException {
        BikeContestProtocol.BikeContestRequest request = ProtocolBuilderUtils.createGetEntriesByRaceRequest(raceID);
        sendRequest(request);

        BikeContestProtocol.BikeContestResponse response = readResponse();
        if (response.getType() == BikeContestProtocol.BikeContestResponse.Type.Error) {
            throw new ContestDataException(ProtocolBuilderUtils.getErrorMessage(response));
        }
        return ProtocolBuilderUtils.getRaceEntries(response);
    }

    @Override
    public Collection<Race> getRacesByEngineCapacity(Integer engineCapacity) throws ContestDataException {
        BikeContestProtocol.BikeContestRequest request
                = ProtocolBuilderUtils.createGetRacesByEngineCapacityRequest(engineCapacity);
        sendRequest(request);

        BikeContestProtocol.BikeContestResponse response = readResponse();
        if (response.getType() == BikeContestProtocol.BikeContestResponse.Type.Error) {
            throw new ContestDataException(ProtocolBuilderUtils.getErrorMessage(response));
        }
        return ProtocolBuilderUtils.getRaces(response);
    }

    @Override
    public Collection<RaceDTO> getRacesWithParticipantCount() throws ContestDataException {
        BikeContestProtocol.BikeContestRequest request
                = ProtocolBuilderUtils.createGetRacesWithParticipantCount();
        sendRequest(request);

        BikeContestProtocol.BikeContestResponse response = readResponse();
        if (response.getType() == BikeContestProtocol.BikeContestResponse.Type.Error) {
            throw new ContestDataException(ProtocolBuilderUtils.getErrorMessage(response));
        }
        return ProtocolBuilderUtils.getRaceDTOs(response);
    }

    @Override
    public void saveRaceEntries(List<RaceEntry> newEntities) throws ContestDataException {
        BikeContestProtocol.BikeContestRequest request
                = ProtocolBuilderUtils.createNewRaceEntriesRequest(newEntities);
        sendRequest(request);
    }

    @Override
    public Collection<Race> getRacesWhereNotRegisteredAndEngineCapacity(Long participantID, Integer engineCapacity) throws ContestDataException {
        BikeContestProtocol.BikeContestRequest request
                = ProtocolBuilderUtils.createGetUnregisteredRacesRequest(participantID, engineCapacity);
        sendRequest(request);

        BikeContestProtocol.BikeContestResponse response = readResponse();
        if (response.getType() == BikeContestProtocol.BikeContestResponse.Type.Error) {
            throw new ContestDataException(ProtocolBuilderUtils.getErrorMessage(response));
        }
        return ProtocolBuilderUtils.getRaces(response);
    }

    @Override
    public Optional<Team> getTeamByName(String teamName) throws ContestDataException {
        BikeContestProtocol.BikeContestRequest request
                = ProtocolBuilderUtils.createGetTeamByNameRequest(teamName);
        sendRequest(request);

        BikeContestProtocol.BikeContestResponse response = readResponse();
        if (response.getType() == BikeContestProtocol.BikeContestResponse.Type.Error) {
            throw new ContestDataException(ProtocolBuilderUtils.getErrorMessage(response));
        }
        return Optional.of(ProtocolBuilderUtils.getTeam(response));
    }

    @Override
    public void saveParticipant(Participant newEntity) throws ContestDataException {
        BikeContestProtocol.BikeContestRequest request
                = ProtocolBuilderUtils.createNewParticipantRequest(newEntity);
        sendRequest(request);
    }

    @Override
    public void saveRace(Race newEntity) throws ContestDataException {
        BikeContestProtocol.BikeContestRequest request
                = ProtocolBuilderUtils.createNewRaceRequest(newEntity);
        sendRequest(request);
    }

    @Override
    public Collection<Team> findAllTeams() throws ContestDataException {
        BikeContestProtocol.BikeContestRequest request = ProtocolBuilderUtils.createGetTeamsRequest();
        sendRequest(request);

        BikeContestProtocol.BikeContestResponse response = readResponse();
        if (response.getType() == BikeContestProtocol.BikeContestResponse.Type.Error) {
            throw new ContestDataException(ProtocolBuilderUtils.getErrorMessage(response));
        }
        return ProtocolBuilderUtils.getTeams(response);
    }

    @Override
    public Collection<Participant> findAllParticipants() throws ContestDataException {
        BikeContestProtocol.BikeContestRequest request = ProtocolBuilderUtils.createGetParticipantsRequest();
        sendRequest(request);

        BikeContestProtocol.BikeContestResponse response = readResponse();
        if (response.getType() == BikeContestProtocol.BikeContestResponse.Type.Error) {
            throw new ContestDataException(ProtocolBuilderUtils.getErrorMessage(response));
        }
        return ProtocolBuilderUtils.getParticipants(response);
    }

    private void sendRequest(BikeContestProtocol.BikeContestRequest request) throws ContestDataException {
        try {
            System.out.println("Sending request..." + request);
            request.writeDelimitedTo(outputStream);
            outputStream.flush();
            System.out.println("Request sent.");
        } catch (IOException ioException) {
            throw new ContestDataException("Error while sending request object: " + ioException);
        }
    }

    private BikeContestProtocol.BikeContestResponse readResponse() throws ContestDataException {
        BikeContestProtocol.BikeContestResponse response = null;

        try {
            response = responses.take();
        } catch (InterruptedException interruptedException) {
            throw new ContestDataException("Error while reading response object: " + interruptedException);
        }
        return response;
    }

    private boolean isUpdateResponse(BikeContestProtocol.BikeContestResponse response) {
        return response.getType() == BikeContestProtocol.BikeContestResponse.Type.NewParticipant
                || response.getType() == BikeContestProtocol.BikeContestResponse.Type.NewRace
                || response.getType() == BikeContestProtocol.BikeContestResponse.Type.NewTeam
                || response.getType() == BikeContestProtocol.BikeContestResponse.Type.UpdatedRaces;
    }

    private void handleUpdate(BikeContestProtocol.BikeContestResponse response) throws ContestDataException {
        switch (response.getType()) {
            case NewParticipant -> client.participantAdded(ProtocolBuilderUtils.getParticipant(response));

            case NewRace -> client.raceAdded(ProtocolBuilderUtils.getRaceDto(response));

            case UpdatedRaces -> client.raceEntriesAdded(ProtocolBuilderUtils.getRaceDTOs(response));
        }
    }

    private void initialiseConnection() {
        try {
            connection = new Socket(host, port);
            outputStream = connection.getOutputStream();
            inputStream = connection.getInputStream();
            finished = false;
            startReader();
        } catch (IOException ioException) {
            ioException.printStackTrace();
        }
    }

    private void closeConnection() {
        finished = true;
        try {
            inputStream.close();
            outputStream.close();
            connection.close();
            client = null;
        } catch (IOException ioException) {
            ioException.printStackTrace();
        }
    }

    private void startReader() {
        Thread threadWorker = new Thread(new ReaderThread());
        threadWorker.start();
    }

    private class ReaderThread implements Runnable {
        public void run() {
            while (!finished) {
                try {
                    BikeContestProtocol.BikeContestResponse response
                            = BikeContestProtocol.BikeContestResponse.parseDelimitedFrom(inputStream);
                    System.out.println("Response received: " + response);

                    if (isUpdateResponse(response)) {
                        handleUpdate(response);
                    } else {
                        try {
                            responses.put(response);
                        } catch (InterruptedException interruptedException) {
                            interruptedException.printStackTrace();
                        }
                    }
                } catch (IOException ioException) {
                    System.out.println("Error while reading: " + ioException.getMessage());
                } catch (ContestDataException contestDataException) {
                    System.out.println(contestDataException.getMessage());
                }
            }
        }
    }
}
